### PR TITLE
Compile .po files into function clauses

### PR DESCRIPTION
**Pull request by whatyouhide**
This has been by far the trickiest thing I had to do, not because it was hard but because I couldn't come up with a clean result that I liked.

One of the things that has given me a push forward is [`filelib:fold_files/5`](http://www.erlang.org/doc/man/filelib.html#fold_files-5), which simplified some of the code I wrote (it both can filter files based on a regex, making it easy to find `.po` files, as well as process directories recursively).

A `__using__` macro has been added to the `Gettext` module in order to compile `.po` files and dynamically generate function clauses (in the caller module) in the form:

```elixir
def lgettext("pt_BR", "default", "Hello world") do
  "Ola mundo"
end
```

In order to test that everything works correctly, I had to add dummy applications inside `test/fixtures/test_applications`. These applications only have an `ebin/` directory (empty except for a `.gitkeep` file for keeping them in git) and translations in them. I'm not sure this is the cleanest approach but I'll wait for more opinions. An alternative solution could be to create the applications dynamically (`File.mkdir_p/1`) and not keeping them into VCS, but it's not very distant to the solution I implemented.

Actually, what I'd like your feedback on the most is the "greater scheme": should I place all the compilation logic inside the `Gettext` module? Alternatively, I could create a `Compiler` module somewhere which would take care of compiling the `.po` files to quoted function clauses to be injected anywhere.

Well, I'm looking forward to your opinions :smile:. Thanks!

**PS** docs are missing, let's see where this goes before adding them.

### CODE REVIEW

**Comment by ericmj** [comment]
You can have a single full application test that calls `use Gettext` and the rest can simply call `Module.eval_quoted __MODULE__, Gettext.compile_po_files(translations_dir)` since that function is public anyway.

**Comment by whatyouhide** [comment]
@ericmj hey, thanks for the feedback. I thought about directly testing `Gettext.compile_po_files/1` and evaluating the quoted result, but this duplicates some logic in the tests. For example, `Gettext` will define [a catchall clause](https://github.com/whatyouhide/gettext/commit/0eaf06340a79ff5f2a042a941ded6a6292eed568#diff-d5c8bd0265515a35a3b4b7d08bb2f6e8R16) into the module that called `use` and we would have to reproduce that manually or to test every functionality related to that catchall clause in the single full application test.

Also, calling `Module.eval_quoted/2` could change as a strategy and we would have to update the tests too in that case. Maybe what we really need is what you suggested, which if I'm not wrong comes down to some kind of *integration* test. What do you think?

File: lib/gettext.ex [hunk 1]
```diff
@@ -1,2 +1,56 @@
 defmodule Gettext do
+  alias Gettext.PO.Translation
+
+  @default_priv "priv/gettext"
+
+  @doc false
+  defmacro __using__(opts) do
+    quote bind_quoted: [opts: opts, default_priv: @default_priv] do
```
Comment (josevalim) [inline lib/gettext.ex]: I think you could do this:

```elixir
defmacro __using__(opts) do
  quote do
    @gettext unquote(opts)
    @before_compile Gettext
  end
end

defmacro __before_compile__(env) do
  po = compile_po_files(translations_dir)

  quote do
    unquote(po)
    # default
  end
end
```

Then you don't need eval quoted and it is relatively cleaner.

**Comment by josevalim** [comment]
> In order to test that everything works correctly, I had to add dummy applications inside test/fixtures/test_applications. 

I wouldn't worry about this. Just explicitly pass the priv directory and use :gettext as `otp_app`. Something like: `use Gettext, otp_app: :gettext, priv: "test/fixtures"`

**Comment by whatyouhide** [comment]
@josevalim right, nice! Is there any way I can get around the fact that the `test/` directory isn't brought into the `Application.app_dir(:gettext)` directory? Should I put the translations at the top level, i.e., something like `gettext/integration/translations`?

**Comment by josevalim** [comment]
Ah, right. We will need to go with your solution then. Be sure to include
an empty .keep file in the ebin directory.

-- 


*José Valim*
www.plataformatec.com.br
Skype: jv.ptec
Founder and Lead Developer

**Comment by whatyouhide** [comment]
Ok, I've added docs to the `Gettext` module so we should be ready to merge this and move on.

File: test/gettext_test.exs [hunk 2]
```diff
@@ -1,8 +1,45 @@
...
-  use ExUnit.Case
-  doctest Gettext
+  use ExUnit.Case, async: true
 
-  test "the truth" do
-    assert 1 + 1 == 2
+  # Dynamically load dummy apps which just have an `ebin/` directory and
+  # translations.
+  for app <- ["default_priv", "custom_priv"] do
+    [__DIR__, "fixtures", "test_applications", app, "ebin"]
```
Comment (josevalim) [inline test/gettext_test.exs]: You need just one application! Then you can have different privs in there, no?

File: lib/gettext.ex [hunk 3]
```diff
@@ -1,2 +1,107 @@
...
+      def lgettext(_, _, msg) do
+        {:default, msg}
+      end
+    end
+  end
+
+  # Made public so that it can be called inside the quoted contents of the
+  # `__using__` macro.
+  @doc false
+  def compile_po_files(dir) do
```
Comment (josevalim) [inline lib/gettext.ex]: This can be private now.

Comment (whatyouhide) [inline lib/gettext.ex]: Good catch!

**Comment by whatyouhide** [comment]
Ok @josevalim, made the changes you suggested. Thanks for the feedback!

**Comment by josevalim** [comment]
:shipit:

**Comment by whatyouhide** [comment]
:dash:

**Pull request merged**